### PR TITLE
[chore][pkg/stanza] Speed up file deduplication in finder

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:** <Describe what has changed.>
For large numbers of files, the logic that deduplicates the filenames between matches is costly. This is mainly due to the O(n^2) deduping algorithm used. If we instead use a map (as a hashset), we can make this ~O(n).

This PR speeds up the deduplication logic, as well as adds a benchmark for a case where the filelog receiver is polling many files at once.

**Testing:** <Describe what testing was performed and which tests were added.>

Running the added benchmark and comparing with benchstat, we can see a large increase in speed for the large number of files case (10000 monitored files), at the cost of a very slight increase in memory usage:
```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/matcher/internal/finder
cpu: Apple M3 Pro
                │    old.txt    │               new.txt               │
                │    sec/op     │    sec/op     vs base               │
Find10kFiles-12   198.636m ± 6%   8.696m ± 16%  -95.62% (p=0.002 n=6)

                │   old.txt    │              new.txt               │
                │     B/op     │     B/op      vs base              │
Find10kFiles-12   5.416Mi ± 0%   5.581Mi ± 0%  +3.04% (p=0.002 n=6)

                │   old.txt   │              new.txt              │
                │  allocs/op  │  allocs/op   vs base              │
Find10kFiles-12   80.06k ± 0%   80.25k ± 0%  +0.23% (p=0.002 n=6)
```

### CODE REVIEW

File: pkg/stanza/fileconsumer/matcher/internal/finder/finder.go [hunk 1]
```diff
@@ -40,15 +41,14 @@ func FindFiles(includes []string, excludes []string) ([]string, error) {
...
-					continue INCLUDE
-				}
-			}
-
-			all = append(all, match)
+			allSet[match] = struct{}{}
 		}
 	}
 
+	all := make([]string, 0, len(allSet))
```
Comment (MovieStoreGuy) [inline pkg/stanza/fileconsumer/matcher/internal/finder/finder.go:48]: You can use `maps.Keys` here instead to simplify the work being done here.

File: pkg/stanza/fileconsumer/matcher/internal/finder/finder_test.go [hunk 2]
```diff
@@ -251,7 +252,31 @@ func TestFindFilesWithIOErrors(t *testing.T) {
...
+
+	includeGlobs := []string{
+		filepath.Join(tmpDir, "log-*.log"),
+	}
+
+	excludeGlobs := []string{}
+
+	b.ResetTimer()
+	for range b.N {
+		_, _ = FindFiles(includeGlobs, excludeGlobs)
```
Comment (MovieStoreGuy) [inline pkg/stanza/fileconsumer/matcher/internal/finder/finder_test.go:280]: This has a high likelihood of being optimised out by the golang compiler, I suggest assigning the value to outside the scope of the loop.

https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

**Review by MovieStoreGuy — COMMENTED** [review]
Couple of pointers, most need to fix up the benchmark to make sure it doesn't get optimised out.

**Review by djaglowski — APPROVED** [review]
Thanks @BinaryFissionGames

File: pkg/stanza/fileconsumer/matcher/internal/finder/finder_test.go (hunk 2 above)
Comment (BinaryFissionGames) [reply pkg/stanza/fileconsumer/matcher/internal/finder/finder_test.go]: Thanks for the tip, I updated to write to a package level variable and reran the benchmarks. Luckily it doesn't look like it was optimized out here.

**Pull request merged**